src/core/fem/src/general/utils/4C_fem_general_extract_values.hpp

```diff
@@ -14,34 +14,112 @@
+   * @return std::array<T, num_entries> Extracted values from the global vector
+   */
+  template <unsigned num_entries, typename T, std::ranges::sized_range Range>
+  std::array<T, num_entries> extract_values_as_array(
+      const Core::LinAlg::Vector<T>& global, const Range& global_ids)
+  {
+    FOUR_C_ASSERT_ALWAYS(global_ids.size() == num_entries, "Number of entries does not match");
+
+    std::array<T, num_entries> local;
+    std::transform(global_ids.begin(), global_ids.end(), local.begin(),
+        [&](int global_id)
+        {
+          const int local_id = global.Map().LID(global_id);
+          FOUR_C_ASSERT_ALWAYS(local_id >= 0,
+              "Proc %d: Cannot find gid=%d in Core::LinAlg::Vector<double>",
+              Core::Communication::my_mpi_rank(global.Comm()), global_id);
+          return global[local_id];
+        });
+
+    return local;
+  }
 
-  Extracts lm.size() values from a distributed epetra vector and stores them into local.
-  this is NOT a parallel method, meaning that all values to be extracted on a processor
-  must be present in global on that specific processor. This usually means that global
-  has to be in column map style.
+  /*!
+   * @brief Extract a subset of local values from Core::LinAlg::Vector<T>
+   *
+   * @tparam T datatype
+   * @param global (in) : Global values
+   * @param lm (in) : Global ids of the values to extract
+   * @return std::vector<T> Extracted values from the global vector
+   */
+  template <typename T, std::ranges::sized_range Range>
+  std::vector<T> extract_values(const Core::LinAlg::Vector<T>& global, const Range& global_ids)
+  {
+    std::vector<T> local(global_ids.size());
+    std::transform(global_ids.begin(), global_ids.end(), local.begin(),
+        [&](int global_id)
+        {
+          const int local_id = global.Map().LID(global_id);
+          FOUR_C_ASSERT_ALWAYS(local_id >= 0,
+              "Proc %d: Cannot find gid=%d in Core::LinAlg::Vector<double>",
+              Core::Communication::my_mpi_rank(global.Comm()), global_id);
+          return global[local_id];
+        });
+
+    return local;
+  }
 
-  \param global (in): global distributed vector with values to be extracted
-  \param local (out): vector or matrix holding values extracted from global
-  \param lm     (in): vector containing global ids to be extracted. Size of lm
-                      determines number of values to be extracted.
-  */
-  void extract_my_values(const Core::LinAlg::Vector<double>& global, std::vector<double>& local,
-      const std::vector<int>& lm);
+  /*!
+   * @brief Legacy function to extract local values from a global vector.
+   *
+   * @note Use @p Core::FE::extract_values() instead
+   *
+   * @param global
+   * @param local
+   * @param lm
+   */
+  inline void extract_my_values(const Core::LinAlg::Vector<double>& global,
+      std::vector<double>& local, const std::vector<int>& lm)
+  {
+    local = extract_values(global, lm);
```

Comment:
I would implement these the other way round. Now I have to make a copy even when I use this signature that could reuse memory.